firehose-writer/src/main/java/com/expedia/www/haystack/pipes/firehoseWriter/FirehoseAsyncHandler.java

```diff
@@ -0,0 +1,99 @@
+ *       See the License for the specific language governing permissions and
+ *       limitations under the License.
+ *
+ */
+package com.expedia.www.haystack.pipes.firehoseWriter;
+
+import com.amazonaws.handlers.AsyncHandler;
+import com.amazonaws.services.kinesisfirehose.model.PutRecordBatchRequest;
+import com.amazonaws.services.kinesisfirehose.model.PutRecordBatchResult;
+import com.amazonaws.services.kinesisfirehose.model.Record;
+import com.netflix.servo.monitor.Counter;
+import com.netflix.servo.monitor.Stopwatch;
+import com.netflix.servo.util.VisibleForTesting;
+
+import java.util.List;
+import java.util.concurrent.Semaphore;
+
+class FirehoseAsyncHandler implements AsyncHandler<PutRecordBatchRequest, PutRecordBatchResult> {
+    @VisibleForTesting
+    static final String RESULT_NULL = "PutRecordBatchResult is null; retrying %d recordList; retryCount=%d";
+
+    @VisibleForTesting
+    final Stopwatch stopwatch;
+    @VisibleForTesting
+    final PutRecordBatchRequest putRecordBatchRequest;
+    @VisibleForTesting
+    final int sleepMillis;
+    @VisibleForTesting
+    final int retryCount;
+    @VisibleForTesting
+    final List<Record> recordList;
+    @VisibleForTesting
+    final RetryCalculator retryCalculator;
+    @VisibleForTesting
+    final Sleeper sleeper;
+    @VisibleForTesting
+    final S3Sender s3Sender;
+    @VisibleForTesting
+    final Semaphore parallelismSemaphore;
+    @VisibleForTesting
+    final Counter throttledCounter;
+    @VisibleForTesting
+    final FailedRecordExtractor failedRecordExtractor;
+
+    FirehoseAsyncHandler(S3Sender s3Sender,
+                         Stopwatch stopwatch,
+                         PutRecordBatchRequest putRecordBatchRequest,
+                         int sleepMillis,
+                         int retryCount,
+                         List<Record> recordList,
+                         RetryCalculator retryCalculator,
+                         Sleeper sleeper,
+                         Semaphore parallelismSemaphore,
+                         Counter throttledCounter,
+                         FailedRecordExtractor failedRecordExtractor) {
+        this.stopwatch = stopwatch;
+        this.putRecordBatchRequest = putRecordBatchRequest;
+        this.sleepMillis = sleepMillis;
+        this.retryCount = retryCount;
+        this.recordList = recordList;
+        this.retryCalculator = retryCalculator;
+        this.sleeper = sleeper;
+        this.s3Sender = s3Sender;
+        this.parallelismSemaphore = parallelismSemaphore;
+        this.throttledCounter = throttledCounter;
+        this.failedRecordExtractor = failedRecordExtractor;
+    }
+
+    @Override
+    public void onError(Exception exception) {
+        s3Sender.onFirehoseCallback(stopwatch, putRecordBatchRequest, null, sleepMillis, retryCount, exception);
+        s3Sender.sendRecordsToS3(recordList, retryCalculator, sleeper, retryCount + 1, parallelismSemaphore);
+    }
```

Comment:
If a specific request is never succeeding (because of bad data rather than a temporary issue) how would be the semaphore be released?